tracing-core/src/callsite.rs

```diff
@@ -169,3 +187,159 @@ impl Hash for Identifier {
         (self.0 as *const dyn Callsite).hash(state)
     }
 }
+
+// ===== impl Registration =====
+
+impl<T> Registration<T> {
+    /// Construct a new `Registration` from some `&'static dyn Callsite`
+    pub const fn new(callsite: T) -> Self {
+        Self {
+            callsite,
+            next: AtomicPtr::new(ptr::null_mut()),
+        }
+    }
+}
+
+impl fmt::Debug for Registration {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Registration").finish()
+    }
+}
+
+// ===== impl LinkedList =====
+
+/// An intrusive atomic push only linked-list.
+struct LinkedList {
+    head: AtomicPtr<Registration>,
+}
+
+impl LinkedList {
+    fn new() -> Self {
+        LinkedList {
+            head: AtomicPtr::new(ptr::null_mut()),
+        }
+    }
+
+    fn for_each(&self, mut f: impl FnMut(&'static dyn Callsite)) {
+        let mut head = self.head.load(Ordering::Acquire);
+
+        while !head.is_null() {
+            let reg = unsafe { &*head };
```

Comment:
TIOLI: it might be a little nicer to express this using [`<*mut T>::as_ref](https://doc.rust-lang.org/std/primitive.pointer.html#method.as_ref) rather than null-checking and then dereferencing. Although the null check means that you will logically never dereference a null pointer, it might be a bit nicer looking to never dereference a pointer which, in the type system, *could* be null:
```suggestion
        while let Some(reg) = unsafe { head.as_ref() } {
```